feat(command): rename variable parts

zardoy opened the pull request.

zardoy committed:
renameVariableParts

zardoy committed:
remove extension diffs

zardoy committed:
almost done, buggy a bit

zardoy commented:
TODO:
- has bug with deleting part
- impl up and down commands (do nothing if editingIndex isnt undefined)

zardoy committed:
ci: use npx as node is preinstalled

zardoy committed:
remove not implemented preserveCamelCase from features

zardoy committed:
ci: ensure type-safety, also duplicate command on main

zardoy committed:
fix eslint & provider error

Ilanaya committed:
small refactor: improve readablility

zardoy commented on src/features/renameVariableParts.ts:
```diff
@@ -81,18 +87,19 @@ export const registerRenameVariableParts = () => {
-                ]
-                quickPick.value = label
-            } else {
+            if (editingIndex) {
```
@ILScc for now, CI is failing (spoiler because of prettier). Also I love refactors of any scale, but this one introduced a bug. As you can see it was checked against undefined on line above as index can be `0`, which is common case.
Because of this change when, editing index 0, user need to double press enter before going out of the so called edit mode (because it would set to -1 on first call). I recommend using [P42](https://p42.ai/documentation/code-assist/#branching) for this:
1. [Invert condition](https://p42.ai/documentation/code-assist/invert-condition) (or Flip if/else from abrocadabra)
2. [Remove unnecessary else](https://p42.ai/documentation/code-assist/remove-unnecessary-else)

Note that all these tools also support Vue.

Of course I could've used linter rule here, but I don't it to be annoying as I experienced a lot of cases, where I couldn't use it because of readability. But it was a good decision in this case.

Ilanaya replied:
Thank for explanation. Fixed it in [679998cb2340b7151863e9477241dfa4c98beb68](https://github.com/zardoy/vscode-experiments/pull/4/commits/679998cb2340b7151863e9477241dfa4c98beb68)

zardoy commented:
@ILScc Check keybindings, everything works fine to me. Try to `alt+k` in shortcuts UI editor and remove all ctrl+l or ctrl+o handlers (you can restore them later in shortcuts.json)

Ilanaya commented:
> @ILScc Check keybindings, everything works fine to me. Try to `alt+k` in shortcuts UI editor and remove all ctrl+l or ctrl+o handlers (you can restore them later in shortcuts.json)

Thanks, this solved the problem. Might worth mentioning in feature description later (if you wanna make such)

Ilanaya committed:
fix refactoring issue

Ilanaya committed:
Back to arrow func

Ilanaya committed:
Update command names

Ilanaya committed:
Implement parts moving

Ilanaya committed:
Update shortcuts

Ilanaya commented on package.json:
```diff
@@ -364,14 +364,14 @@
                 "command": "zardoyExperiments.renameVariablePartsPartMoveDown",
-                "key": "ctrl+k",
-                "mac": "cmd+k",
+                "key": "ctrl+alt+k",
```
Updated this shortcut because `Ctrl+k` is reserved for chords. Moving up shortcut is updated for the sake of consistency

Ilanaya committed:
Resolve prettier

Ilanaya committed:
Resolve linting problems

Ilanaya committed:
Merge branch 'main' into renameVariableParts

zardoy committed:
[skip ci] fix: extension crash on empty changes
fix: make autoRemoveSemicolon disableable

zardoy committed:
Merge branch 'main' into renameVariableParts

zardoy commented:
Ensure that:
1. Editing any part performs casing normalization (it should make first-uppercase if not first)
2. It does the same when moving parts
When I tested it, moving parts didn't work correctly

Ilanaya committed:
Fix incorrect behaviour, make resulting case correctness

Ilanaya commented:
@zardoy  looks like I've managed to deal with moving parts :)
Can you please describe bug you mentioned above - I'll try fixing if it's still relevant
> TODO:
> 
> * has bug with deleting part
> * impl up and down commands (do nothing if editingIndex isnt undefined)

Ilanaya committed:
Merge branch 'main' into renameVariableParts

Ilanaya committed:
Resolve prettier

zardoy committed:
support dash and underscore casings (not mixed)

zardoy committed:
fix mispellings in command names

zardoy commented:
Yes, deleting part normalizes first character, but not editing. Imagine these two cases:

1. Edit part:

`someCoolVariable` -> `someDecentVariable`

When you select `Cool` part and replace it with `decent` you shouldn't care about first character, as you're editing *one part*. It should autocapitalized, if not first.

2. Append or prepend text in some part:

`someCoolVariable` -> `someReallyCoolVariable`

You can select `Cool` part and prepend the text with `really`

Also I checked that it doesn't work with all casings:

```ts
const im_should_not_be_capitalized = 5
// or
const a = {
    'some-prop': 5
}
```

Would be great to support them as well. I *pushed 1 commit* with that support.

zardoy commented:
I think it'd great also to refactor `parts` array to use proxy from *valtio* instead. Example: https://github.com/zardoy/npm-the-fastest/blob/0474087adb0d7070fa3f500368a5be9eefc1ae4c/src/features/statusbar.ts#L2

```ts
const parts = proxy([])
// subscribe and update first-letter uppercase in parts whenever they change
```

Instead of duplicating logic it'd be much cleaner

zardoy committed:
Restore futureFeatures.ts

zardoy committed:
robust case preserving
robust active index preserving
more informative title: displays pos or range
change shortcuts for mac to avoid conflicts

zardoy committed:
implement changing item to lowercase (mergin with previous one)
change shorcut for replace action

zardoy commented:
Now everything works, only two tasks remaining:

- (small) allow to type big/small letter to change isPascalCase (or even refactor to get rid of it)
- when editing: `SomethingElse` -> enter should add `Else` as a part

However it can be merged with these tasks in mind.

zardoy committed:
Merge remote-tracking branch 'origin/main' into renameVariableParts

Ilanaya committed:
Update dep

zardoy committed:
fix linter
up lockfile

zardoy committed:
fix command titles

zardoy committed:
fix pascal editing
